dns_server/src/crawler_p2p/crawler/address_data.rs

```diff
@@ -0,0 +1,326 @@
+    },
+
+    PermanentlyUnreachable {
+        /// Same as above
+        fail_count: u32,
+
+        /// Same as above
+        was_reachable: bool,
+
+        /// At which time the address would be removed from memory
+        erase_after: Duration,
+    },
+}
+
+/// Additional state of a potential node address
+pub struct AddressData {
+    /// Connection state
+    pub state: AddressState,
+
+    /// Whether the address was added from the command line
+    pub user_added: ConstValue<bool>,
+}
+
+impl AddressState {
+    fn fail_count(&self) -> u32 {
+        match self {
+            AddressState::Connecting {
+                fail_count,
+                was_reachable: _,
+            } => *fail_count,
+            AddressState::Connected {} => 0,
+            AddressState::Disconnecting {
+                fail_count,
+                was_reachable: _,
+            } => *fail_count,
+            AddressState::Disconnected {
+                fail_count,
+                was_reachable: _,
+                disconnected_at: _,
+            } => *fail_count,
+            AddressState::PermanentlyUnreachable {
+                fail_count,
+                erase_after: _,
+                was_reachable: _,
+            } => *fail_count,
+        }
+    }
+
+    fn was_reachable(&self) -> bool {
+        match self {
+            AddressState::Connecting {
+                fail_count: _,
+                was_reachable,
+            } => *was_reachable,
+            AddressState::Connected {} => true,
+            AddressState::Disconnecting {
+                fail_count: _,
+                was_reachable,
+            } => *was_reachable,
+            AddressState::Disconnected {
+                fail_count: _,
+                was_reachable,
+                disconnected_at: _,
+            } => *was_reachable,
+            AddressState::PermanentlyUnreachable {
+                erase_after: _,
+                was_reachable,
+                fail_count: _,
+            } => *was_reachable,
+        }
+    }
+
+    /// Whether the address is currently recognized as reachable (available from DNS)
+    pub fn is_reachable(&self) -> bool {
+        match self {
+            AddressState::Connecting {
+                fail_count: _,
+                was_reachable: _,
+            } => false,
+            AddressState::Connected {} => true,
+            AddressState::Disconnecting {
+                fail_count: _,
+                was_reachable: _,
+            } => false,
+            AddressState::Disconnected {
+                fail_count: _,
+                was_reachable: _,
+                disconnected_at: _,
+            } => false,
+            AddressState::PermanentlyUnreachable {
+                fail_count: _,
+                was_reachable: _,
+                erase_after: _,
+            } => false,
+        }
+    }
+
+    /// Whether to retain the address between node restarts (store in DB)
+    pub fn is_persistent(&self) -> bool {
```

Comment:
I don't think this is valid. The crawler has no idea about persistence.

Reply:
It's only used in `CrawlerManager`.